x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/mapper/DataTierFieldMapper.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.datastreams.mapper;
+
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.mapper.ConstantFieldType;
+import org.elasticsearch.index.mapper.MetadataFieldMapper;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.search.lookup.SearchLookup;
+import org.elasticsearch.xpack.cluster.routing.allocation.DataTierAllocationDecider;
+
+import java.util.Collections;
+import java.util.function.Supplier;
+
+public class DataTierFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_tier";
+
+    public static final String CONTENT_TYPE = "_tier";
+
+    public static final TypeParser PARSER = new FixedTypeParser(c -> new DataTierFieldMapper());
+
+    static final class DataTierFieldType extends ConstantFieldType {
+
+        static final DataTierFieldType INSTANCE = new DataTierFieldType();
+
+        private DataTierFieldType() {
+            super(NAME, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        // Commented out below because triggered some false expectations in SQL - their tests assume members of the keyword family return
+        // aggs? The expected vs actual table results showed an unexpected `_tier` column with a null value.
+        //
+
+        // @Override
+        // public String familyTypeName() {
+        // return KeywordFieldMapper.CONTENT_TYPE;
+        // }
+
+        @Override
+        protected boolean matches(String pattern, boolean caseInsensitive, SearchExecutionContext context) {
+            if (caseInsensitive) {
+                pattern = Strings.toLowercaseAscii(pattern);
+            }
+            String tierPreference = context.getIndexSettings().getSettings().get(DataTierAllocationDecider.INDEX_ROUTING_PREFER, "not_set");
```

Comment:
I wonder if we should only match against the last element of the setting. By default, cold tier allocation will be "hot,warm,cold", which I think makes specifying the search criteria to hit a tier more complex than necessary, especially if you have some allocated to "cold" and some to "hot,warm,cold".

Reply:
Trouble is that's potentially changing the logic of how queries like `term` or `terms` work on keyword fields.
If we go down that route it might be more logical to think of it as a range query (e.g.search for  >= warm) but that again messes with the normal logic of sort orders on keyword fields which is just alphabetical on the text of the keywords

Reply:
I think it's fine to consider only the preferred tier. The list in the preference is ordered so `data_warm,data_hot` means that the index is in the warm tier but it can allocate in the hot tier if there are no nodes in the warm tier. The fact that the index is allocated on the fallback doesn't change the status of the index. It's still an index in the warm tier waiting to be allocated there. 
So imo, the `_tier` metadata field could be the first preference in the list only. That's the tier that represents the data.

Reply:
Ah. I misread your question @henningandersen 
I thought your complaint was on the query side i.e users needing to pass a list in a `terms` query e.g.

     "terms": { "_tier":  ["data_hot", "data_warm"] } }

I guess the problem is with the value held in the setting which can be a comma delimited string rather than a single value?

Reply:
Exactly, see also: https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-migrate.html#ilm-migrate

Reply:
If the "first preference" in the list is what we query, presumably that should be the only value we return when we add support for aggregations?

Reply:
+1, it's the logical tier where the data belongs